src/expr/parser.rs

```diff
@@ -1,60 +1,123 @@
+
+use expr::Expr;
+use expr::BinOp;
 
 use nom::digit;
 
+use std::mem;
 use std::str;
 use std::str::FromStr;
 
-named!(parens<i64>, delimited!(
+named!(parens<Expr>, delimited!(
     char!('('),
     expr,
     char!(')')
   )
 );
 
-named!(i64_literal<i64>,
-  map_res!(
+named!(parse_isize<isize>,
     map_res!(
-      digit,
-      str::from_utf8
-    ),
-    FromStr::from_str
-  )
+      map_res!(digit, str::from_utf8),
+      FromStr::from_str
+    )
+);
+
+named!(isize_literal<Expr>,
+    map!(parse_isize, |x| Expr::Literal(x))
 );
 
-named!(factor<i64>,
+named!(factor<Expr>,
   alt!(
-    i64_literal
+    isize_literal
   | parens
   )
 );
 
 // we define acc as mutable to update its value whenever a new term is found
-named!(term <i64>,
+named!(term <Expr>,
   chain!(
     mut acc: factor  ~
-             many0!(
-               alt!(
-                 tap!(mul: preceded!(tag!("*"), factor) => acc = acc * mul) |
-                 tap!(div: preceded!(tag!("/"), factor) => acc = acc / div)
-               )
-             ),
+      many0!(
+        alt!(
+          map!(preceded!(tag!("*"), term), |prod: Expr| {
+            match acc {
+              Expr::Literal(num) => {
+                acc = Expr::BinOp(Box::new(Expr::Literal(num)),
+                                  BinOp::Times, Box::new(prod));
+              },
+              Expr::BinOp(ref mut x, ref mut op, ref mut y) => {
+                // swap out each of acc's values
+                let inner_x = mem::replace(x, Box::new(prod));
+                let inner_op = mem::replace(op, BinOp::Times);
+                let inner_y = mem::replace(y, Box::new(Expr::Literal(0)));
+                mem::replace(y, Box::new(
+                        Expr::BinOp(inner_x, inner_op, inner_y)));
+              }
+            }
+          }) |
+          map!(preceded!(tag!("/"), term), |div: Expr| {
+            match acc {
+              Expr::Literal(num) => {
+                acc = Expr::BinOp(Box::new(Expr::Literal(num)), BinOp::Over,
+                                  Box::new(div));
+              },
+              Expr::BinOp(ref mut x, ref mut op, ref mut y) => {
+                let inner_x = mem::replace(x, Box::new(Expr::Literal(0)));
+                let inner_op = mem::replace(op, BinOp::Over);
+                let inner_y = mem::replace(y, Box::new(div));
+                mem::replace(x, Box::new(
+                        Expr::BinOp(inner_x, inner_op, inner_y)));
+              }
+            }
+          })
+        )
+      ),
     || { return acc }
   )
 );
 
-named!(pub expr <i64>,
+named!(pub expr <Expr>,
   chain!(
     mut acc: term  ~
-             many0!(
-               alt!(
-                 tap!(add: preceded!(tag!("+"), term) => acc = acc + add) |
-                 tap!(sub: preceded!(tag!("-"), term) => acc = acc - sub)
-               )
-             ),
+      many0!(
+        alt!(
+          map!(preceded!(tag!("+"), term), |add: Expr| {
```

Comment:
Okay, so you swtiched to `map!`, which means that you're taking ownership of `add`. So far so good!

Now your central problem is how to avoid moving `acc` into the closure, because if you do this, then `acc` will not be usable in the future.